torch/csrc/api/include/torch/data/datasets/chunk.h

```diff
@@ -0,0 +1,455 @@
+  virtual ChunkType read_chunk(size_t chunk_index) = 0;
+
+  /// Returns the number of chunks available in this reader.
+  virtual size_t chunk_count() = 0;
+
+  /// This will clear any internal state associate with this reader.
+  virtual void reset() = 0;
+};
+
+namespace detail {
+/// BatchDataBuffer manages a queue of UnwrappedBatchData. After a new chunk is
+/// loaded, BatchDataBuffer splits it into small batches and push them into the
+/// queue. When get_batch is called from data loader, it pops cached batches and
+/// return. If the cache is empty, it either waits to load more chunks or return
+/// null if all chunks are loaded.
+template <
+    typename UnwrappedBatch = std::vector<Example<>>,
+    typename ExampleSampler = samplers::RandomSampler>
+class BatchDataBuffer {
+ public:
+  using UnwrappedBatchType = UnwrappedBatch;
+  using BatchType = torch::optional<UnwrappedBatchType>;
+  using BatchRequestType = typename ExampleSampler::BatchRequestType;
+
+  BatchDataBuffer(
+      size_t num_chunks,
+      size_t batch_size,
+      ExampleSampler example_sampler,
+      bool ignore_empty_chunk,
+      size_t queue_capacity)
+      : remaining_chunk_count_(num_chunks),
+        batch_size_(batch_size),
+        example_sampler_(std::move(example_sampler)),
+        ignore_empty_chunk_(ignore_empty_chunk),
+        queue_capacity_(queue_capacity) {}
+
+  /// Return batch data from the queue. Called from the ChunkDataset main
+  /// thread.
+  BatchType get_batch() {
+    std::unique_lock<std::mutex> lock(queue_mutex_);
+    cv_read_.wait(lock, [this] {
+      // wait till there is available data in the queue or if all chunks are
+      // loaded (i.e. the dataset is exhausted for this epoch)
+      return (
+          this->total_example_count_in_queue_ >= batch_size_ ||
+          this->remaining_chunk_count_ == 0);
+    });
+    if (batch_queue_.empty()) {
+      lock.unlock();
+      AT_ASSERT(remaining_chunk_count_ == 0);
+
+      // All batches have been retrieved. Return an empty batch.
+      return nullopt;
+    }
+
+    auto batch = batch_queue_.front();
+    batch_queue_.pop();
+    if (batch.exception) {
+      throw WorkerException(batch.exception);
+    }
+
+    total_example_count_in_queue_ -= batch.batch_data.size();
+    lock.unlock();
+    cv_write_.notify_all();
+
+    return batch.batch_data;
+  }
+
+  // skip one chunk
+  void skip_chunk() {
+    std::unique_lock<std::mutex> lock(queue_mutex_);
+    AT_ASSERT(remaining_chunk_count_ > 0);
+    remaining_chunk_count_--;
+    lock.unlock();
+    cv_read_.notify_all();
+  }
+
+  /// Push preloaded chunks to batch queue. Called from the ChunkDataset worker
+  /// threads.
+  void add_chunk_data(UnwrappedBatchType data) {
+    std::unique_lock<std::mutex> lock(queue_mutex_);
+    cv_write_.wait(lock, [this] {
+      // stop loading if we have preloaded enough data.
+      return this->total_example_count_in_queue_ < this->queue_capacity_ || stop_;
```

Comment:
FWIW `queue_capacity_` isn't strictly enforced, so it's slightly misleading. In reality the upper bound is `queue_capacity_` + _the maximum number of examples per chunk_. You would need to suspend this thread when writing if you wanted to enforce this.

Reply:
Also, the total bound on the number of elements in memory will be:
`queue_capacity_ + num_loaders * max_examples_per_chunk`.

Reply:
This is true. The relaxation of the fixed size make this implementation simpler.  The queue capacity does not need to be big and mostly not the main concern, but the max_examples_per_chunk. This is what we expect the readers to handle, keeping the chunks comparatively smaller. For example, we can suggest something like 256MB per chunk.

Reply:
Fair enough.